guides/bulk.md

```diff
@@ -52,18 +52,25 @@ client.bulk({
 ```
 As you can see, each bulk operation is comprised of two objects. The first object contains the operation type and the target document's `_index` and `_id`. The second object contains the document's data. As a result, the body of the request above contains six objects for three index actions.
 
-Alternatively, the `bulk` method can accept an array of hashes where each hash represents a single operation. The following code is equivalent to the previous example:
+Alternatively, the `bulk` method can accept an array of hashes where hashes are in pairs, except for delete which is singular. In each pair, the first item contains the action like `create` and, `_index` and `id` specify which index and document id the action has to be performed on. The second item in the pair is the document itself.
 
 ```javascript
 client.bulk({
-  body: [
-    { index: { _index: movies, _id: 1, data: { title: 'Beauty and the Beast', year: 1991 }  } },
-    { index: { _index: movies, _id: 2, data: { title: 'Beauty and the Beast - Live Action', year: 2017 } } },
-    { index: { _index: books, _id: 1, data: { title: 'The Lion King', year: 1994 } } }
-  ]
-}).then((response) => {
-  console.log(response);
-});
+    index: movies,
+    body: [
+      { create: { _id: 1} },
+      { title: 'Beauty and the Beast 1', year: 2050 },
+      { delete: { _id: 1 } },
+      { create: { _id: 2 } },
+      { title: 'Beauty and the Beast 2', year: 2051 },
+      { create: {} },
+      { title: 'Beauty and the Beast 2', year: 2051 },
+      { create: { _index: books } },
+      { title: '2012', year: 2012 },
+    ]
+  }).then((response) => {
+    console.log(response.body.items);
+  });
 ```
 
 We will use this format for the rest of the examples in this guide.
```

Comment:
Looks like this part is not applicable to JavaScript. There's only way to format the body for `client.bulk` in JS unlike in Ruby.

Reply:
As in we can remove line 55 to 76

Reply:
Look like we have `client.helpers.bulk` that allows another format if you wanna look into that.

Reply:
> As in we can remove line 55 to 76

I don't follow exactly. Is that not how client.bulk works? Because I have used that same method across the document.

Reply:
I think you should write a working sample, and include it along with these changes so we can be sure @AbhinavGarg90.

Reply:
@dblock The code snippets  in each guide are supposed to be working examples that can be run from start to finish, in order, without throwing errors. It's the reason @AbhinavGarg90 is fixing it right now because the current snippets throw errors.


@AbhinavGarg90 OpenSearch only accepts 1 body format for the bulk action which is an array of operation/document pair like so:
```ruby
 [
    { create: { _index: movies, _id: 1 } },
    { title: 'Beauty and the Beast', year: 1991 },
    { delete: { _index: movies, _id: 1 } },
    { index:  { _index: books, _id: 2 } },
    { title: 'The Lion King', year: 1994 }
 ]
 ```
 
 However, the implementation of this action on Ruby also accepts another format that's simpler:
 ```ruby
[
    { create: { _index: movies, _id: 1, data: { title: 'Beauty and the Beast', year: 1991 }  } },
    { delete: { _index: movies, _id: 1},
    { index:  { _index: books, _id: 1, data: { title: 'The Lion King', year: 1994 } } }
]
```
This format is then translated to the former one before sending it is serialized and sent to the server.
`client.bulk` in javascript doesn't support this format. However looks like `client.helpers.bulk` also provides similar syntactical sugar. That's why I think you should look into that and add it to this guide as well, because users looking to use `client.bulk` will likely wanna know about `client.helpers.bulk` as well.

Reply:
> The code snippets in each guide are supposed to be working examples 

I'm just saying that we need to commit the working samples and run them if we want them to be ... working.

Reply:
> Look like we have `client.helpers.bulk` that allows another format if you wanna look into that.

I can work on adding this, can you explain a bit on how this function is supposed to be called? I looked through the code, and I'm not completely sure.

Reply:
I don't have good sample code, but https://opensearch.org/docs/2.2/clients/javascript/helpers/ seems to be quite thorough, and I found a few people reporting bugs with otherwise working examples, e.g [here](https://github.com/opensearch-project/opensearch-js/issues/192).

Reply:
Great, I'll look through this and add it to the pull request. Once I'm done with that I'll try to look into the other issue as well.

Reply:
```js
const { Client } = require('@opensearch-project/opensearch')
const documents = require('./docs.json')

const client = new Client({ ... })

const result = await client.helpers.bulk({
  datasource: documents,
  onDocument (doc) {
    return {
      index: { _index: 'example-index' }
    }
  }
})

console.log(result)
```

Looking at the documentation, it doesn't specify the shape of the input document. I assume it has key and value, but I cannot seem to locate any documentation with further detail about this.

Reply:
You mean https://opensearch.org/docs/latest/api-reference/document-apis/bulk/? It's pretty complete (bulk API accepts line-delimited JSON).

Reply:
> You mean https://opensearch.org/docs/latest/api-reference/document-apis/bulk/? It's pretty complete (bulk API accepts line-delimited JSON).

Thank you for pointing out where this is. Should I link this in bulk.md as well?

Reply:
How does line-delimited JSON work? I tried running a sample using the available documentation, it doesn't seem to be working.

Here are the code snippets I am working with:

```js
const { Client } = require('@opensearch-project/opensearch');
const documents = require('./docs.json')

try {
  const client = new Client({
    node: 'https://admin:admin@localhost:9200',
    ssl: { rejectUnauthorized: false }
  });
  client.helpers.bulk({
    datasource: documents,
    onDocument (doc) {
      return {
        index: { _index: 'example-index' }
      }
    }
  }).then((response) => {
    console.log(response)
  }).catch((error) => {
    console.error(error);
  })
} catch(e) {
  console.log("error", e)
}

```
docs.json
```json
{ "delete": { "_index": "movies", "_id": "tt2229499" } }
{ "index": { "_index": "movies", "_id": "tt1979320" } }
{ "title": "Rush", "year": 2013 }
{ "create": { "_index": "movies", "_id": "tt1392214" } }
{ "title": "Prisoners", "year": 2013 }
{ "update": { "_index": "movies", "_id": "tt0816711" } }
{ "doc" : { "title": "World War Z" } }
```

error:
```sh
Unexpected non-whitespace character after JSON at position 57
```

Reply:
@AbhinavGarg90 Sorry for the very late response. I've been out of commission for the last few weeks.

The datasource is an iterable and the simplest iterable in this case is an array of objects, where each object is an OpenSearch doc:
```javascript
  const docs = [{ title: 'Beauty and the Beast 1', year: 2050 },
                { title: 'Beauty and the Beast 2', year: 2051 }]
```
Then you can pass it to the helper function:
```javascript
  let result = await client.helpers.bulk({ datasource: docs,
    onDocument(doc) {
      return { index: { _index: movies} }
    }
  });

  console.log(result);
```
And you should get this in the console:
```
{
  total: 2,
  failed: 0,
  retry: 0,
  successful: 2,
  noop: 0,
  time: 744,
  bytes: 150,
  aborted: false
}
```

Basically the `client.helpers.bulk` method, on top of implementing some batching logic, also acts as syntactical sugar for `client.bulk`. That is, the user only have to provide the documents in their original forms (an array of objects in the above example) and the action to apply (index in this example) to apply to the documents.